app/models/key.rb

```diff
@@ -1,4 +1,85 @@
+require 'securerandom'
+
 class Key < ActiveRecord::Base
+  has_many :logs
   belongs_to :place
-  has_and_belongs_to_many :clients
+  belongs_to :guest
+  before_create :set_secret_url
+
+  def self.get_available_key keys
+    keys.each do |key|
+      client_key = key.client_keys.find_by(requested: true, used_at: nil)
+      master_key = key.client_keys.find_by(requested: true, unlimited_access: true)
+      return client_key if client_key
+      return master_key if master_key
+    end
+    nil
+  end
+
+  def cleaner_time(start_or_end)
+    self.send(start_or_end).strftime("%I:%M %p %a %b %e, %Y")
+  end
+
+  def place_and_time
+    "#{self.place.nickname} between #{self.start_date} and #{self.end_date}"
+  end
+
+  def name
+    key = Key.find_by(id: self.key_id)
+    client = Client.find_by(id: self.client_id)
+    time = time_until(key.start_date)
+    if Time.now.utc > key.end_date
+      "#{client.name} had access to #{key.place.nickname}"
+    elsif Time.now.utc > key.start_date && Time.now.utc < key.end_date
+      "#{client.name} currently has access to #{key.place.nickname}"
+    else
+    "#{client.name} will have access to #{key.place.nickname} in #{time}"
+    end
+  end
+
+  def get_start_date_string
+    start_date.strftime("%Y/%m/%d %H:%M:%S")
+  end
+
+  def is_available?
+    current_time = Time.now.in_time_zone("Eastern Time (US & Canada)")
+    key.start_date <= current_time && key.end_date >= current_time && !used_at
+  end
+
+  private
+
+  def set_secret_url
+    self.secret_url = SecureRandom.urlsafe_base64
+  end
+
+  def time_until(time_stamp)
```

Comment:
This should probably be a class method / not on this model. It's not reading anything from the instance and is just doing some time-related jiggery pokery.